flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/BigQuerySchemaProviderImpl.java

```diff
@@ -0,0 +1,347 @@
+        // These are newly added.
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.timeMillis().getName(), FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.timeMicros().getName(), FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.localTimestampMillis().getName(),
+                FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.localTimestampMicros().getName(),
+                FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put("geography_wkt", FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put("Json", FieldDescriptorProto.Type.TYPE_STRING);
+    }
+
+    // --------------- Obtain AvroSchema from TableSchema -----------------
+    /**
+     * Function to convert TableSchema to Avro Schema.
+     *
+     * @param tableSchema A {@link TableSchema} object to cast to {@link Schema}.
+     * @return Converted Avro Schema
+     */
+    private static Schema getAvroSchema(TableSchema tableSchema) {
+        return SchemaTransform.toGenericAvroSchema("root", tableSchema.getFields());
+    }
+
+    // --------------- Obtain Descriptor Proto from Avro Schema  ---------------
+    /**
+     * Obtains a Descriptor Proto by obtaining Descriptor Proto field by field.
+     *
+     * <p>Iterates over Avro Schema to obtain FieldDescriptorProto for it.
+     *
+     * @param schema Avro Schema, for which descriptor is needed.
+     * @return DescriptorProto describing the Schema.
+     */
+    private static DescriptorProto getDescriptorSchemaFromAvroSchema(Schema schema) {
+        // Iterate over each table field and add them to the schema.
+        Preconditions.checkState(!schema.getFields().isEmpty());
+
+        DescriptorProto.Builder descriptorBuilder = DescriptorProto.newBuilder();
+        // Create a unique name for the descriptor ('-' characters cannot be used).
+        // Replace with "_" and prepend "D".
+        descriptorBuilder.setName(BigQueryUtils.bqSanitizedRandomUUIDForDescriptor());
+        int i = 1;
+        for (Schema.Field field : schema.getFields()) {
+            fieldDescriptorFromSchemaField(field, i++, descriptorBuilder);
+        }
+        return descriptorBuilder.build();
+    }
+
+    /**
+     * Function to obtain the FieldDescriptorProto from a AvroSchemaField and then append it to
+     * DescriptorProto builder.
+     *
+     * @param field {@link Schema.Field} object to obtain the {@link FieldDescriptorProto} from.
+     * @param fieldNumber index at which the obtained {@link FieldDescriptorProto} is appended in
+     *     the Descriptor.
+     * @param descriptorProtoBuilder {@link DescriptorProto.Builder} object to add the obtained
+     *     {@link FieldDescriptorProto} to.
+     */
+    private static void fieldDescriptorFromSchemaField(
```

Comment:
this function does a lot of stuff and i'm having a hard time following. i think it'd be helpful to clearly describe the behavior in the function comment. for example: what conversion is it doing? when does it throw exceptions? etc

another thing that might help is separating this out into some well-named helper functions

Reply:
You are right. Fixed.

Reply:
yes that helps a lot! you can resolve this convo